File: algorithms/2_medium/L/00817_linked_list_components/table.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
#include<threads.h>

#define MAX_ALLOWED_NUM 10000

thread_local static uint8_t table[MAX_ALLOWED_NUM + 1];
thread_local static uint8_t stamp = 0;

int numComponents(struct ListNode *restrict head, const int *restrict nums, const int len) {
    if (stamp == UINT8_MAX) {
        stamp = 0;
        for (size_t i = 0; i <= MAX_ALLOWED_NUM; i += 1) {
            table[i] = 0;
        }
    }
    stamp += 1;
    for (int i = 0; i < len; i += 1) {
        if (nums[i] < 0 || nums[i] > MAX_ALLOWED_NUM) {
            return -1;
        }
        table[nums[i]] = stamp;
    }

    bool is_prev_valid = false;
    int cnt = 0;
    struct ListNode *restrict curr = head;

    while (curr != NULL) {
        const int val = curr->val;
        curr = curr->next;

        if (val < 0 || val > MAX_ALLOWED_NUM) {
            return -1;
        }

        if (table[val] == stamp) {
            cnt += is_prev_valid ? 0 : 1;
            is_prev_valid = true;
        } else {
            is_prev_valid = false;
        }
    }

    return cnt;
}

File: algorithms/2_medium/L/00817_linked_list_components/table.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
use std::cell::UnsafeCell;

const MAX_ALLOWED_NUM: i32 = 10_000;
const TABLE_LEN: usize = (MAX_ALLOWED_NUM + 1) as usize;

thread_local! { 
    static TABLE: UnsafeCell<[u8; TABLE_LEN]> = UnsafeCell::new([0; TABLE_LEN]);
    static STAMP: UnsafeCell<u8> = UnsafeCell::new(0);
}

impl Solution {
    pub fn num_components(head: Option<Box<ListNode>>, nums: Vec<i32>) -> i32 {
        let table = unsafe { &mut *(TABLE.with(|t| t.get())) };
        let stamp_ref = unsafe { &mut *(STAMP.with(|s| s.get())) };

        if *stamp_ref == u8::MAX {
            table.fill(0);
            *stamp_ref = 1;
        } else {
            *stamp_ref += 1;
        }

        let stamp = *stamp_ref;
        for &num in &nums {
            table[num as usize] = stamp;
        }
        let mut is_prev_valid = false;
        let mut cnt = 0_i32;
        let mut curr_opt = head;

        while let Some(mut curr) = curr_opt {
            curr_opt = curr.next.take();
            if table[curr.val as usize] == stamp {
                cnt += if is_prev_valid { 0 } else { 1 };
                is_prev_valid = true;
            } else {
                is_prev_valid = false;
            }
        }

        cnt
    }
}
